File: python/examples/backend/providers/amazon_bedrock.py
import asyncio
from typing import Final

from pydantic import BaseModel, Field

from beeai_framework.adapters.amazon_bedrock.backend.chat import AmazonBedrockChatModel
from beeai_framework.backend.chat import ChatModel
from beeai_framework.backend.events import ChatModelNewTokenEvent
from beeai_framework.backend.message import UserMessage
from beeai_framework.cancellation import AbortSignal
from beeai_framework.emitter import EventMeta
from beeai_framework.errors import AbortError
from beeai_framework.parsers.field import ParserField
from beeai_framework.parsers.line_prefix import LinePrefixParser, LinePrefixParserNode

# NOTE: See README.md for additional usage notes
MODEL_NAME: Final[str] = "meta.llama3-8b-instruct-v1:0"


async def amazon_bedrock_from_name() -> None:
    llm = ChatModel.from_name(f"amazon_bedrock:{MODEL_NAME}")
    user_message = UserMessage("what states are part of New England?")
    response = await llm.create(messages=[user_message])
    print(response.get_text_content())


async def amazon_bedrock_sync() -> None:
    llm = AmazonBedrockChatModel(MODEL_NAME)
    user_message = UserMessage("what is the capital of Massachusetts?")
    response = await llm.create(messages=[user_message])
    print(response.get_text_content())


async def amazon_bedrock_stream() -> None:
    llm = AmazonBedrockChatModel(MODEL_NAME)
    user_message = UserMessage("How many islands make up the country of Cape Verde?")
    response = await llm.create(messages=[user_message], stream=True)
    print(response.get_text_content())


async def amazon_bedrock_stream_abort() -> None:
    llm = AmazonBedrockChatModel(MODEL_NAME)
    user_message = UserMessage("What is the smallest of the Cape Verde islands?")

    try:
        response = await llm.create(messages=[user_message], stream=True, abort_signal=AbortSignal.timeout(0.5))

        if response is not None:
            print(response.get_text_content())
        else:
            print("No response returned.")
    except AbortError as err:
        print(f"Aborted: {err}")


async def amazon_bedrock_structure() -> None:
    class TestSchema(BaseModel):
        answer: str = Field(description="your final answer")

    llm = AmazonBedrockChatModel(MODEL_NAME)
    user_message = UserMessage("How many islands make up the country of Cape Verde?")
    response = await llm.create_structure(schema=TestSchema, messages=[user_message])
    print(response.object)


async def amazon_bedrock_stream_parser() -> None:
    llm = AmazonBedrockChatModel(MODEL_NAME)

    parser = LinePrefixParser(
        nodes={
            "test": LinePrefixParserNode(
                prefix="Prefix: ",
                field=ParserField.from_type(str),
                is_start=True,
                is_end=True,
            )
        }
    )

    async def on_new_token(data: ChatModelNewTokenEvent, event: EventMeta) -> None:
        await parser.add(chunk=data.value.get_text_content())

    user_message = UserMessage("Produce 3 lines each starting with 'Prefix: ' followed by a sentence and a new line.")
    await llm.create(messages=[user_message], stream=True).observe(
        lambda emitter: emitter.on("new_token", on_new_token)
    )
    result = await parser.end()
    print(result)


async def main() -> None:
    print("*" * 10, "amazon_bedrock_from_name")
    await amazon_bedrock_from_name()
    print("*" * 10, "amazon_bedrock_sync")
    await amazon_bedrock_sync()
    print("*" * 10, "amazon_bedrock_stream")
    await amazon_bedrock_stream()
    print("*" * 10, "amazon_bedrock_stream_abort")
    await amazon_bedrock_stream_abort()
    # NOTE: Disabled by default -- see README for more information
    # print("*" * 10, "amazon_bedrock_structure")
    # await amazon_bedrock_structure()
    print("*" * 10, "amazon_bedrock_stream_parser")
    await amazon_bedrock_stream_parser()


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/examples/backend/providers/amazon-bedrock.ts
import "dotenv/config";
import { AmazonBedrockChatModel } from "beeai-framework/adapters/amazon-bedrock/backend/chat";
import "dotenv/config.js";
import { ToolMessage, UserMessage } from "beeai-framework/backend/message";
import { ChatModel } from "beeai-framework/backend/chat";
import { z } from "zod";
import { ChatModelError } from "beeai-framework/backend/errors";
import { OpenMeteoTool } from "beeai-framework/tools/weather/openMeteo";

const llm = new AmazonBedrockChatModel(
  "amazon.titan-text-lite-v1",
  // {},
  // {
  //   accessKeyId: "AWS_ACCESS_KEY_ID",
  //   secretAccessKey: "AWS_SECRET_ACCESS_KEY",
  //   region: "AWS_REGION",
  //   sessionToken: "AWS_SESSION_TOKEN",
  // },
);

llm.config({
  parameters: {
    topK: 1,
    temperature: 0,
    topP: 1,
  },
});

async function amazonBedrockFromName() {
  const amazonBedrockLLM = await ChatModel.fromName("amazon-bedrock:amazon.titan-text-lite-v1");
  const response = await amazonBedrockLLM.create({
    messages: [new UserMessage("what states are part of New England?")],
  });
  console.info(response.getTextContent());
}

async function amazonBedrockSync() {
  const response = await llm.create({
    messages: [new UserMessage("what is the capital of Massachusetts?")],
  });
  console.info(response.getTextContent());
}

async function amazonBedrockStream() {
  const response = await llm.create({
    messages: [new UserMessage("How many islands make up the country of Cape Verde?")],
    stream: true,
  });
  console.info(response.getTextContent());
}

async function amazonBedrockAbort() {
  try {
    const response = await llm.create({
      messages: [new UserMessage("What is the smallest of the Cape Verde islands?")],
      stream: true,
      abortSignal: AbortSignal.timeout(1 * 1000),
    });
    console.info(response.getTextContent());
  } catch (err) {
    if (err instanceof ChatModelError) {
      console.error("Aborted", { err });
    }
  }
}

async function amazonBedrockStructure() {
  const response = await llm.createStructure({
    schema: z.object({
      answer: z.string({ description: "your final answer" }),
    }),
    messages: [new UserMessage("How many islands make up the country of Cape Verde?")],
  });
  console.info(response.object);
}

async function amazonBedrockToolCalling() {
  const userMessage = new UserMessage("What is the weather in Boston?");
  const weatherTool = new OpenMeteoTool({ retryOptions: { maxRetries: 3 } });
  const response = await llm.create({ messages: [userMessage], tools: [weatherTool] });
  const toolCallMsg = response.getToolCalls()[0];
  console.debug(JSON.stringify(toolCallMsg));
  const toolResponse = await weatherTool.run(toolCallMsg.args as any);
  const toolResponseMsg = new ToolMessage({
    type: "tool-result",
    result: toolResponse.getTextContent(),
    toolName: toolCallMsg.toolName,
    toolCallId: toolCallMsg.toolCallId,
  });
  console.info(toolResponseMsg.toPlain());
  const finalResponse = await llm.create({ messages: [userMessage, toolResponseMsg], tools: [] });
  console.info(finalResponse.getTextContent());
}

async function amazonBedrockDebug() {
  // Log every request
  llm.emitter.match("*", (value, event) =>
    console.debug(
      `Time: ${event.createdAt.toISOString()}`,
      `Event: ${event.name}`,
      `Data: ${JSON.stringify(value)}`,
    ),
  );

  const response = await llm.create({
    messages: [new UserMessage("Hello world!")],
  });
  console.info(response.messages[0].toPlain());
}

console.info(" amazonBedrockFromName".padStart(25, "*"));
await amazonBedrockFromName();
console.info(" amazonBedrockSync".padStart(25, "*"));
await amazonBedrockSync();
console.info(" amazonBedrockStream".padStart(25, "*"));
await amazonBedrockStream();
console.info(" amazonBedrockAbort".padStart(25, "*"));
await amazonBedrockAbort();
console.info(" amazonBedrockStructure".padStart(25, "*"));
await amazonBedrockStructure();
console.info(" amazonBedrockToolCalling".padStart(25, "*"));
await amazonBedrockToolCalling();
console.info(" amazonBedrockDebug".padStart(25, "*"));
await amazonBedrockDebug();
